### PR TITLE
networkd: allow setting rp_filter for an interface

### PR DESCRIPTION
**Pull request by ssahani**
closes https://github.com/systemd/systemd/issues/26241

### CODE REVIEW

File: src/network/networkd-network-gperf.gperf [hunk 1]
```diff
@@ -140,6 +140,8 @@ Network.PrimarySlave,                        config_parse_bool,
 Network.IPv4ProxyARP,                        config_parse_tristate,                                    0,                             offsetof(Network, proxy_arp)
 Network.ProxyARP,                            config_parse_tristate,                                    0,                             offsetof(Network, proxy_arp)
 Network.IPv6ProxyNDPAddress,                 config_parse_ipv6_proxy_ndp_address,                      0,                             0
+Network.IPv4ReversePathFilter,               config_parse_int,                                         0,                             offsetof(Network, ipv4_rp_filter)
+Network.IPv6ReversePathFilter,               config_parse_int,                                         0,                             offsetof(Network, ipv6_rp_filter)
```
Comment (yuwata) [inline src/network/networkd-network-gperf.gperf]: Please introduce a specific parser for the settings, that should be take one of 'no' (or boolean false), ’strict', or 'loose'.

File: man/systemd.network.xml [hunk 2]
```diff
@@ -776,6 +776,28 @@ Table=1234</programlisting></para>
...
+        <term><varname>IPv4ReversePathFilter=</varname></term>
+        <listitem>
+          <para>Configures IPv4 Reverse Path Filtering. Reverse Path Filtering is used to prevent packets that arrived
+          via one interface from leaving via a different interface. When a machine with reverse path filtering enabled
+          receives a packet, the machine will first check whether the source of the received packet is reachable through
+          the interface it came in. If it is routable through the interface which it came, then the machine will accept the packet.
+          If it is not routable through the interface, which it came, then the machine will drop that packet. Takes one of
+          <literal>0</literal> — No source validation,
+          <literal>1</literal> — Strict mode or
+          <literal>2</literal> — Loose mode. See <ulink url="https://tools.ietf.org/html/rfc1027">RFC 3704</ulink>.
```
Comment (yuwata) [inline man/systemd.network.xml]: Could you briefly explain the difference between the strict and loose mode?

File: man/systemd.network.xml [hunk 3]
```diff
@@ -776,6 +776,28 @@ Table=1234</programlisting></para>
...
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>IPv4ReversePathFilter=</varname></term>
+        <listitem>
+          <para>Configures IPv4 Reverse Path Filtering. Reverse Path Filtering is used to prevent packets that arrived
+          via one interface from leaving via a different interface. When a machine with reverse path filtering enabled
+          receives a packet, the machine will first check whether the source of the received packet is reachable through
+          the interface it came in. If it is routable through the interface which it came, then the machine will accept the packet.
+          If it is not routable through the interface, which it came, then the machine will drop that packet. Takes one of
```
Comment (yuwata) [inline man/systemd.network.xml]: I think it's better to drop the comma before 'which'.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/network/networkd-network-gperf.gperf (hunk 1 above)
Comment (ssahani) [reply src/network/networkd-network-gperf.gperf]: yes right

**Comment by ssahani** [comment]
Updated . thanks for the review !

File: man/systemd.network.xml [hunk 4]
```diff
@@ -776,6 +776,31 @@ Table=1234</programlisting></para>
...
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>IPv4ReversePathFilter=</varname></term>
+        <listitem>
+          <para>Configures IPv4 Reverse Path Filtering. Reverse Path Filtering is used to prevent packets that arrived
+          via one interface from leaving via a different interface. When a machine with reverse path filtering enabled
+          receives a packet, the machine will first check whether the source of the received packet is reachable through
+          the interface it came in. If it is routable through the interface which it came, then the machine will accept the packet.
+          If it is not routable through the interface which it came, then the machine will drop that packet. Takes one of
```
Comment (keszybz) [inline man/systemd.network.xml]: This description uses two terms: "reachable" and "routable". I think it'd be easier to understand if the text was more direct:
> Configure IPv4 Reverse Path Filtering. If enabled, when an IPv4 packet is received, the machine will first check whether the `<emphasis>source</emphasis>` of the packet would be routed through the interface it came in. If there is no route to the source on that interface, the machine will drop the packet.

File: man/systemd.network.xml [hunk 5]
```diff
@@ -776,6 +776,31 @@ Table=1234</programlisting></para>
...
+          <literal>no</literal> No source validation.
+          <literal>strict</literal> In <literal>strict</literal> mode each incoming packet is tested against the FIB and
+          if the incoming interface is not the best reverse path, the packet check will fail. By default failed packets are discarded.
+          <literal>loose</literal> In <literal>loose</literal> mode each incoming packet's source address is tested against the FIB.
```
Comment (keszybz) [inline man/systemd.network.xml]: Something is wrong with capitalization here. I'd expect something like `<literal>no</literal> — no source validation`.

File: man/systemd.network.xml [hunk 6]
```diff
@@ -776,6 +776,31 @@ Table=1234</programlisting></para>
...
+          The packet is dropped only if the source address is not reachable via any interface on that router.
+          See <ulink url="https://tools.ietf.org/html/rfc1027">RFC 3704</ulink>.
+          When unset, the kernel's default will be used.</para>
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>IPv6ReversePathFilter=</varname></term>
+        <listitem>
+          <para>Configures IPv6 Reverse path filtering. As in <varname>IPv4ReversePathFilter=</varname>.</para>
```
Comment (keszybz) [inline man/systemd.network.xml]: `"Configures IPv6 Reverse path filtering. Same as <varname>IPv4ReversePathFilter=</varname>, but for IPv6 packets."`

**Review by keszybz — COMMENTED** [review]
LGTM, except for docs formatting/wording.

**Comment by keszybz** [comment]
This seems very low risk, we can merge this before the release.

**Comment by ssahani** [comment]
updated thanks for the review

File: man/systemd.network.xml [hunk 7]
```diff
@@ -776,6 +776,29 @@ Table=1234</programlisting></para>
         </listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>IPv4ReversePathFilter=</varname></term>
+        <listitem>
+          <para>Configure IPv4 Reverse Path Filtering. If enabled, when an IPv4 packet is received, the machine will first check
+          whether the <emphasis>source</emphasis> of the packet would be routed through the interface it came in. If there is no
+          route to the source on that interface, the machine will drop the packet. Takes one of
```
Comment (yuwata) [inline man/systemd.network.xml:784]:
```
Takes one of <literal>no</literal>, <literal>strict</literal>, or <literal>loose</literal>. When <literal>no</literal>,
no source validation will be done. When <literal>strict</literal>, each incoming packet is tested...
When <literal>loose</literal>, each incoming packet's source addres is tested...
```

**Review by yuwata — CHANGES_REQUESTED** [review]
Could you add a simple test in test/test-network ?

**Comment by ssahani** [comment]
> Could you add a simple test in test/test-network ?

I figure ipv6 rp_filter does not work like ip4. Hence dropped. Thanks for the review

**Comment by keszybz** [comment]
LGTM, but let's merge this after the next release. It seems that it might be more complicated than it seemed at first, and we don't want to introduce new code at this point.

File: src/network/networkd-network.h [hunk 8]
```diff
@@ -302,6 +302,7 @@ struct Network {
         int ipv4_route_localnet;
         int ipv6_dad_transmits;
         int ipv6_hop_limit;
+        int ipv4_rp_filter;
```
Comment (yuwata) [inline src/network/networkd-network.h]: `int` -> `IPReversePathFilter`

File: src/network/networkd-network.c [hunk 9]
```diff
@@ -474,6 +474,7 @@ int network_load_one(Manager *manager, OrderedHashmap **networks, const char *fi
                 .ipv6_hop_limit = -1,
                 .ipv6_proxy_ndp = -1,
                 .proxy_arp = -1,
+                .ipv4_rp_filter = -1,
```
Comment (yuwata) [inline src/network/networkd-network.c]: `-1` -> `_IP_REVERSE_PATH_FILTER_INVALID`

**Review by yuwata — CHANGES_REQUESTED** [review]

**Comment by ssahani** [comment]
updated thanks for the review !

**Review by bluca — APPROVED** [review]

**Pull request merged**